sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala

```diff
@@ -193,6 +193,16 @@ case object SinglePartition extends Partitioning {
   }
 }
 
+/**
+ * Represents a partitioning where rows are only serialized/deserialized locally. The number
+ * of partitions are not changed and also the distribution of rows. This is mainly used to
+ * obtain some statistics of map tasks such as number of outputs.
+ */
+case class LocalPartitioning(orgPartition: Partitioning, numPartitions: Int) extends Partitioning {
```

Comment:
This might be expensive as soon as we hit the sort based shuffle. Perhaps we should carve out some specialized shuffle writing path for this at some point. You basically only need to write to a single file and your done.

Reply:
Not sure if I understand correctly. We explicitly specify this `LocalPartitioning` when doing global limit and submit a map stage using this partitioner. Why we possibly hit a sort based shuffle?

> You basically only need to write to a single file and your done.
I think this is what we want. I specify the same partition numbers for `LocalPartitioning`  as its child RDD and the rows in a partition all have the same partition id when using `LocalPartitioning`. Doesn't it make it to write to a single file?

Reply:
It will hit the sort based shuffle path as soon as the `numPartitions` > 200 right? The problem is not that it will end up in the same shuffle file (it will), the (small) problem is that the sort based shuffle buffers rows and tries to sort them before writing them out. It is just a lot of unneeded complexity.

Reply:
Ah. I see. Thanks for the clarifying. I agree that we might need to have specialized shuffle writing path at some point. Currently I think when we hit the sort based shuffle, this should not be worse performance than previous global limit operation. If you agree, I'd like to put it to some follow-ups.

Reply:
One more thing, can you make LocalRelation use `orgPartition.numPartitions` instead of adding the it as a separate field?

Reply:
Ok.

Reply:
As you see that this causes test failure. It is because `child.outputPartitioning` can be an `UnknownPartitioning(0)`.  To avoid using a separate field, I make childRDD as `LocalRelation`'s field instead of `orgPartition`.